NorthstarDedicatedTest/masterserver.cpp

```diff
@@ -927,6 +954,99 @@ void MasterServerManager::RemoveSelfFromServerList()
 	requestThread.detach();
 }
 
+void MasterServerManager::GetPing(RemoteServerInfo* server)
+{
+	if (server == NULL) {
+		return;
+	}
+
+	std::thread requestThread([this, server]()
+		{
+			if (!server->ipSet)
+				return;
+
+			server->pingPending = true;
+			try {
+				spdlog::info("Attempting to get ping of server \"{}\"", server->name);
+				// Declare and initialize variables
+				HANDLE hIcmpFile;
+				unsigned long ipaddr = INADDR_NONE;
+				DWORD dwRetVal = 0;
+				DWORD dwError = 0;
+				char SendData[32] = "Data Buffer";
+				LPVOID ReplyBuffer = NULL;
+				DWORD ReplySize = 0;
+				DWORD Timeout = 5000;
+				ipaddr = server->ip.S_un.S_addr;
+				if (ipaddr == INADDR_NONE || ipaddr == 0) {
+					server->pingPending = true;
+					return;
+				}
+				hIcmpFile = IcmpCreateFile();
+				if (hIcmpFile == INVALID_HANDLE_VALUE) {
+					spdlog::error(fmt::format("Encountered an error pinging server \"{}\". Error: Unable to open handle.", server->name));
+					server->pingPending = false;
+					return;
+				}
+				ReplySize = sizeof(ICMP_ECHO_REPLY) + sizeof(SendData);
+				ReplyBuffer = (VOID*)malloc(ReplySize);
+				if (ReplyBuffer == NULL) {
+					spdlog::error(fmt::format("Encountered an error pinging server \"{}\". Error: Unable to allocate memory", server->name));
+					server->pingPending = false;
+					return;
+				}
+				dwRetVal = IcmpSendEcho2(hIcmpFile, NULL, NULL, NULL,
```

Comment:
Many machines will have ICMP packets blocked via firewall eg all windows machines by default which will make this ineffective. Additionally ICMP is low priority and will not be an accurate indicator of latency. This should be done by sending a query packet to the UDP game port(s) and measuring the time until a response is received, and ideally average a few.

Reply:
The problem with pinging the UDP port is that it exposes which port the server is using and negates any reason for people to set up servers on non-default ports as it loses all security value.
From my testing about 40-50% of servers had ICMP allowed. Server owners could also just be asked to allow ICMP if they want their server’s ping showing in the server list.
It also actually seems like it’s a relatively accurate indicator by chance, deviating ~10ms due to the time lost by ICMP’s low priority approximately lining up with the game server’s internal latency.

Reply:
I’m not sure I understand your first point. Unless the game traffic is proxied through the master which is highly inefficient and a bad idea, the client is still going to be transmitting and receiving packets directly to/from the game ports which is extremely trivial to figure out with any packet capturing software. To my knowledge in context where latency is extremely important there is no way to truly obfuscate remote endpoint connections so any attempt is going to be trivial to bypass.

Reply:
It exposes the port to people not trying to join the server, making it easy for them to be attacked

Reply:
Once again a malicious actor can very easily determine the game port. Using ICMP is not going to do anything other slow them down for maybe 5 seconds.

Reply:
Yeah but I also wrote the code to do ICMP and i had looked into making it do a TCP/UDP ping which is way harder. But god was it a pain in the arse to program it even for ICMP, I'd rather just stick with what I've got.